### PR TITLE
Add SelectXXXIterator separate to WhereSelectXXXIterator

### PR DESCRIPTION
**Pull request by JonHanna**
`WhereSelectXXXIterator` classes can no longer have a null predicate, so that case need no longer be checked for. (Add readonly and assertions to ensure this assumption is true.)

Add IList<T>-based Select specialised class, on same basis; increases cases covered by optimised `ToArray()`, also reduces number of branches non-ilist sources (likely most common case) go
through though at cost of one more for array or list.

IArrayProvider<TElement>.ToArray() will now never return null, so don't test for the case that it may have.

### CODE REVIEW

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 1]
```diff
@@ -46,11 +47,16 @@ private static IEnumerable<TSource> WhereIterator<TSource>(IEnumerable<TSource>
...
-            List<TSource> list = source as List<TSource>;
-            if (list != null) return new WhereSelectListIterator<TSource, TResult>(list, null, selector);
-            return new WhereSelectEnumerableIterator<TSource, TResult>(source, null, selector);
+            IList<TSource> ilist = source as IList<TSource>;
+            if (ilist != null)
+            {
+                TSource[] array = source as TSource[];
+                if (array != null) return new SelectArrayIterator<TSource, TResult>(array, selector);
+                List<TSource> list = source as List<TSource>;
+                if (list != null) return new SelectListIterator<TSource, TResult>(list, selector);
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Why the specialization for ```List<T>```?  I assume this was done purely for performance, e.g. avoiding the interface calls?  If so, what kinds of gains do you see?

Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: I see, we have the ```List<T>``` special case for Where, but not the ```IList<T>`` case.  Why was it important to add that one for Select?

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 2]
```diff
@@ -400,16 +419,153 @@ public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> s
...
+                return false;
+            }
+
+            public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector)
+            {
+                return new WhereSelectListIterator<TSource, TResult2>(_source, _predicate, CombineSelectors(_selector, selector));
+            }
+        }
+
+        internal sealed class SelectEnumerableIterator<TSource, TResult> : Iterator<TResult>
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: What's the value of writing SelectEnumerableIterator out by hand rather than just implementing it as:
```foreach (var item in source) yield return selector(source);```?  It doesn't implement ToArray, and no other code tries to use its selector, so I'm not clear what value this provides.  Is it purely for the Select(...).Select(...) case in order to be able to use CombineSelectors?

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 3]
```diff
@@ -447,14 +603,10 @@ public override bool MoveNext()
                         state = 2;
                         goto case 2;
                     case 2:
-                        while (_enumerator.MoveNext())
+                        if (_enumerator.MoveNext())
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: I was expecting to see here code that used ```List<T>``` indexer rather than getting an enumerator and using MoveNext/Current.  Is that not good for some reason?

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 4]
```diff
@@ -351,22 +372,20 @@ public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> s
             {
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: We used to have test coverage for this method.  With these changes, this method is now uncovered by tests.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 5]
```diff
@@ -400,16 +419,153 @@ public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> s
...
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Ditto.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 6]
```diff
@@ -464,18 +616,76 @@ public override bool MoveNext()
 
             public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector)
             {
-                return new WhereSelectListIterator<TSource, TResult2>(_source, _predicate, CombineSelectors(_selector, selector));
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Ditto.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 7]
```diff
@@ -464,18 +616,76 @@ public override bool MoveNext()
...
+                }
+                base.Dispose();
+            }
+
+            public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> selector)
+            {
+                return new SelectIListIterator<TSource, TResult2>(_source, CombineSelectors(_selector, selector));
+            }
+
+            public TResult[] ToArray()
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Please ensure we have tests for this; it doesn't appear to be exercised by any of the tests.

**Comment by stephentoub** [comment]
cc: @VSadov 

Thanks, @JonHanna. Overall it looks good to me.  I'd like to see the test gaps addressed.  It'd also be helpful to understand what the performance improvements look like.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: It's the IList case that's truly new here, with the others being derived from the previous code. I would probably have just done the IList case if coming at a tabla rasa, or maybe just the IList and the array, but I'm loath to remove a previous optimisation unless I'm convinced it's actually making things slower (or if it hits a buggy case, obviously).
The IList case can benefit in terms of ToArray (and there's scope for further work on the cases where an iterator can know its size in advance) with select, so I added that, but it can't with Where, so I left that as is.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 2 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: CombineSelectors is its only real benefit, though there's a tiny benefit when it's subsequently used in a Where.
It was though pretty quick to write by hand given `WhereSelectEnumerableIterator` as a starting point, and there's greater consistency between the different cases.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 3 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: That would change the behaviour if the source list is altered during enumeration. (I've never been convinced the dangers of that are such as to require an outright ban, but it's way past the point of debating that). The test `Select_SourceListGetsModifiedDuringIteration_ExceptionIsPropagated` checks on this.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 2 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: > It was though pretty quick to write by hand given WhereSelectEnumerableIterator as a starting point, and there's greater consistency between the different cases.

in general I'm less concerned in these cases about how hard it was to write, and more concerned with an increase in the amount of code that needs to be maintained.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 3 above)
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: That makes sense.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 2 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Easy to write doesn't always correspond with easy to maintain, but it can. I see the main advantage being the consistency, which could help maintenance if a future change can be consistently applied.
Or not, since that's a bit crystal-ball.
In the balance, there is also CombineSelectors.

Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: And for CombineSelectors, this was an optimization that already existed for this case, and by splitting Select out of the WhereSelect, it kind of forces the hand. Ok.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 3 above)
Comment (justinvp) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Should `SelectListIterator.ToArray` use `List<T>.Enumerator` as well, for the same reason?

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: No need. List<T>.Enumerator throws to prohibit code like:

```C#
foreach(var item in list)
{
  // alter list here and don't jump out of the loop
}

And `Select_SourceListGetsModifiedDuringIteration_ExceptionIsPropagated` insists that we throw on:

```C#
foreach(var item in list.Select(λ))
{
  // alter list here and don't jump out of the loop
}

`ToArray` doesn't allow for that possibility, so iterating is a valid optimisation.

(There's the potential for problems if the source list to be changed during the call by another thread, but that involves getting non-threadsafe behaviour from methods that don't claim to be threadsafe, and could fail either way).

**Comment by JonHanna** [comment]
I won't get to those tests for a couple of days, but agree they should certainly be in there before this is merged.

**Comment by stephentoub** [comment]
> I won't get to those tests for a couple of days, but agree they should certainly be in there before this is merged.

@JonHanna, just checking in... you still working on this?  No rush, just verifying.

**Comment by JonHanna** [comment]
@stephentoub I will be, but apart from looking in on notifications, I won't be doing anything for at least another few days, as I'm busy with several matters.

**Comment by stephentoub** [comment]
Ok, thanks for the update. Sounds good.

**Comment by JonHanna** [comment]
@stephentoub I've added tests to cover those cases, copying those for the sequence `.Where(…).Select(…)` to deal with `.Where(…).Select(…).Select(…)`.

I wrote something that compares the before and after of these changes at https://github.com/JonHanna/EnumerablePerformanceComparison. It has an `Enumerable` class in the namespace `System.LinqTest.New` and another in `System.LinqTest.Old` and switching which namespace is imported into the test class between runs changes which is tested. I ran it through Windows DNX on a core i7 with the CPU set to have a minimum and maximum of 90% to reduce the impact of Turbo Boost on the results.

Combined results can be seen at https://github.com/JonHanna/EnumerablePerformanceComparison/raw/master/compare.ods

The tests are run for both `ToArray` and a `Consume` that simply spins through the enumerable as quickly as possible. The times include the `Select` and `Where` operations. The Where predicates are pretty artificial (one that is always false, one that is always true, and one that toggles a captured boolean and returns that). Different element types are tried, for arrays, lists, non-list enumerables, readonly collections and a cut down copy of `List<T>` (these last two giving two different `IList` implementations that don't have their own specific paths). The sizes of the collections are a few multiples of 10 and of 2 in case binary-round sizes gave particularly good or particularly bad results.

The biggest change was, as expected, for `ToArray()` on a `Select` without `Where` where the source is an `IList` that is not a `List` or an array. This case is improved by around 60%, which I have to say is even better than I'd thought it would be.

The other cases where much less pronounced, often being within each others margin of error (showing an improvement in one run, and a deterioration in another). However, they average out over runs at around 5% improvement. Cases where the source was an array were the least improved (quite a few runs showing deterioration, and averaging only slightly improved) but non-list enumerables (probably the most common case, especially if there are other linq operations prior to this) were the most improved, around 6-7% for `ToArray()` and approaching 10% for consuming the enumerable.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 8]
```diff
@@ -135,12 +147,14 @@ void IEnumerator.Reset()
 
         internal class WhereEnumerableIterator<TSource> : Iterator<TSource>
```
Comment (Maxwe11) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: `private` ?

**Comment by JonHanna** [comment]
Oh, one case I new would be affected negatively was that of an array or list source to a `Where` where the source has zero elements. It's affected because there's no longer the optimisation for that situation in 780ef34 which I'd introduced because at the time it was cheap, but with these changes it no longer is. Since that optimisation had only been introduced recently, and since the case is a cheap one anyway (being only hit when there are zero results after all) I don't think that case going back to previous timings (perhaps even slightly better) is a big concern.

File: src/System.Linq/src/System/Linq/Enumerable.cs (hunk 8 above)
Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: It probably should be. This class has been `internal` since it was brought into the repository, and I made the comparable classes introduced here, like `SelectEnumerableIterator`, `internal` to match that. If there's no good reason for these to be `internal` then I think they should all be made `private` to give a better ability to eyeball that they're correct. While I made a few fields `readonly` for similar ability to assure oneself of correctness at a glance, that was only because I'd changed the degree that invariant was depended on (if `_predicate` changed, part of my logic was flawed).
In other words, I think you're right, but I think I shouldn't change it right here were it's not relevant.

**Comment by stephentoub** [comment]
Thanks, @JonHanna.

@VSadov, thoughts?

**Comment by VSadov** [comment]
LGTM

**Comment by stephentoub** [comment]
@JonHanna, can you rebase this, at least to get rid of the merge commit in there?

**Comment by JonHanna** [comment]
@stephentoub squashed.

File: src/System.Linq/src/System/Linq/Enumerable.cs [hunk 9]
```diff
@@ -400,16 +419,153 @@ public override IEnumerable<TResult2> Select<TResult2>(Func<TResult, TResult2> s
...
+            {
+                return new SelectEnumerableIterator<TSource, TResult>(_source, _selector);
+            }
 
+            public override void Dispose()
+            {
+                if (_enumerator != null)
+                {
+                    _enumerator.Dispose();
+                    _enumerator = null;
```
Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Hmm. I just realized, this is going to cause a NullReferenceException in MoveNext where one didn't previously exist, right?

Comment (JonHanna) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: No, the base Dispose is called next, which guarantees the state checked by MoveNext will ignore this field and return false.

Comment (stephentoub) [inline src/System.Linq/src/System/Linq/Enumerable.cs]: Ah, you're right, thanks.

**Comment by stephentoub** [comment]
LGTM

**Comment by VSadov** [comment]
LGTM
